### PR TITLE
Add support for reading protected (RC4 encrypted) .xls files

### PR DESCRIPTION
**Pull request by trvrnrth**
These changes make it possible to read in .xls files with protection enabled.

The decryption functions and objects implemented here are are based on the source of Spreadsheet-ParseExcel:
http://search.cpan.org/~jmcnamara/Spreadsheet-ParseExcel/

### CODE REVIEW

**Comment by goosehub** [comment]
I found this pull request through https://stackoverflow.com/questions/19655690/how-to-read-password-protected-excel-files-using-phpexcel. I'm not clear on the result. I see this referenced commit, https://github.com/PHPOffice/PHPExcel/commit/c7a6431e099067becda33986ee664fe52ea10e33, appears to exist in master, but I'm not sure how to use the repo to pass a password to a password protected xls file.

In short, does this repo support opening password protected xls files, either in Master or in this branch? If so, where is an example, documentation on this feature. If not, how should I start on forking this repo so I can build the feature and possibly make a pull request back to the repo.

**Comment by trvrnrth** [comment]
The functionality did make it into master yes. I never implemented the ability to use anything other than the default password for "read only" protected sheets (VelvetSweatshop) but I imagine it'd be possible to pass a different password in via the reader as an option or something and then make use of that.

**Comment by goosehub** [comment]
@trvrnrth Oh good. I'll have no problem changing the password in Excel5.php, as every file I'll receive will have the same password.

Could you provide documentation or an example of how to open password protected files? This is what I have now, even after changing the password.

    $excel_object = PHPExcel_IOFactory::load($xlsx_filepath);
    $sheet_data = $excel_object->getActiveSheet()->toArray(null, false, true, true);

I'm getting

> Unexpected file pass record length

**Comment by trvrnrth** [comment]
@goosehub You know, I honestly couldn't tell you anymore it was so long ago now. The project I used the library one is no longer maintained by me. What I can tell you is that at the time this worked well for the read-only excel 5 files we had to deal with but we never had a use case for handling those with an actual password. I borrowed the approach from the perl Spreadsheet-ParseExcel lib so it might be worth you seeing if that can read your file first.

**Comment by goosehub** [comment]
Yeah, it's not even getting to the `$this->verifyPassword('VelvetSweatshop', ...` (which I've updated to my password) line because it's failing at

        $length = self::getInt2d($this->data, $this->pos + 2);

        if ($length != 54) {
            throw new PHPExcel_Reader_Exception('Unexpected file pass record length');
        }

And if I remove that exception, it fails at the verifyPassword line and throws this exception

        throw new PHPExcel_Reader_Exception('Decryption password incorrect');

And removing that exception just causes PHP errors.

I'm guessing getting it to work will be non trivial. I'm waiting for a potential other option to get the data we need, but if that falls through, I'll probably have to make this work somehow.

**Comment by PowerKiKi** [comment]
If you happen to create a PR for that, please target PhpSpreadsheet, instead of PHPExcel.

**Comment by goosehub** [comment]
@PowerKiKi so I fork PhpSpreadsheet and build the functionality into that? I assume both repos are able to parse xls files?

**Comment by PowerKiKi** [comment]
Yep, it would be best. PhpSpreadsheet is PHPExcel renamed, it's mostly the same code base, so same features. You should also cover you work with unit tests, so that the feature doesn't break again in the future.
